app/uk/gov/hmrc/apinotificationqueue/controller/QueueController.scala

```diff
@@ -25,55 +25,67 @@ import play.api.http.HttpEntity
 import play.api.libs.json.Json
 import play.api.mvc._
 import uk.gov.hmrc.apinotificationqueue.model.{Notification, Notifications}
-import uk.gov.hmrc.apinotificationqueue.service.QueueService
-import uk.gov.hmrc.http.BadRequestException
+import uk.gov.hmrc.apinotificationqueue.service.{ApiSubscriptionFieldsService, QueueService}
+import uk.gov.hmrc.http.HeaderCarrier
 import uk.gov.hmrc.play.bootstrap.controller.BaseController
 import uk.gov.hmrc.play.http.logging.MdcLoggingExecutionContext._
 
 import scala.concurrent.Future
 
 @Singleton()
-class QueueController @Inject()(queueService: QueueService, idGenerator: NotificationIdGenerator) extends BaseController {
+class QueueController @Inject()(queueService: QueueService, fieldsService: ApiSubscriptionFieldsService, idGenerator: NotificationIdGenerator) extends BaseController {
 
+  private val SUBSCRIPTION_FIELD_HEADER_NAME = "api-subscription-fields-id"
   private val CLIENT_ID_HEADER_NAME = "X-Client-ID"
-  private val CLIENT_ID_REQUIRED_ERROR = new BadRequestException(s"$CLIENT_ID_HEADER_NAME required.")
+
+  private val MISSING_CLIENT_ID_ERROR = s"$CLIENT_ID_HEADER_NAME required."
+  private val MISSING_BODY_ERROR = s"Body required."
 
   def save() = Action.async {
     implicit request => {
       val headers = request.headers
-      val clientId = headers.get(CLIENT_ID_HEADER_NAME).getOrElse(throw CLIENT_ID_REQUIRED_ERROR)
-      val notificationId = idGenerator.generateId()
-      queueService.save(
-        clientId,
-        Notification(
-          notificationId,
-          headers.remove(CLIENT_ID_HEADER_NAME).toSimpleMap,
-          request.body.asXml.getOrElse(throw new BadRequestException("No body included.")).toString,
-          DateTime.now()
-        )
-      )
-      Future.successful(Result(
-        header = ResponseHeader(status = CREATED, headers = Map(LOCATION -> routes.QueueController.get(notificationId).url)),
-        body = HttpEntity.NoEntity
-      ))
+      getClientId(headers).flatMap(_.fold(Future.successful(BadRequest(MISSING_CLIENT_ID_ERROR))) {
+        clientId =>
+          request.body.asXml.fold(Future.successful(BadRequest(MISSING_BODY_ERROR))) { body =>
+            queueService.save(
+              clientId,
+              Notification(
+                idGenerator.generateId(),
+                headers.remove(CLIENT_ID_HEADER_NAME, SUBSCRIPTION_FIELD_HEADER_NAME).toSimpleMap,
+                body.toString(),
+                DateTime.now()
+              )
+            ).map(notification =>
+              Result(ResponseHeader(CREATED, Map(LOCATION -> routes.QueueController.get(notification.notificationId).url)), HttpEntity.NoEntity))
+          }
+
+      })
     }
   }
 
+  private def getClientId(headers: Headers)(implicit hc: HeaderCarrier): Future[Option[String]] = {
```

Comment:
This private method is called only inside `save() (when you call `POST /queue`).
Is it correct?
Or shall we use it in all the other endpoints?
If the the `X-Client-ID` header is not provided in the other endpoints, shall we check the `api-subscription-fields-id` header first and then retrieve the client id?

Reply:
The requirement is only there for the POST everything else is a failure case if not. Possible to move it across them all if needed

Reply:
Fair enough.